subprojects/core/src/test/groovy/org/gradle/internal/progress/BuildProgressLoggerTest.groovy

```diff
@@ -172,4 +175,26 @@ class BuildProgressLoggerTest extends Specification {
         1 * provider.start("Configure projects", _) >> confProgress
         1 * progress1.completed()
     }
+
+    def "avoids trying to complete the same logger twice in multi-threaded environment"() {
```

Comment:
We should probably break up this tests into two different test cases:

1. Just test that the correct interactions are invoked (your first `when`/`then`) independent of a multi-threaded scenario.
2. A test case that tests that concurrent access works properly. At the moment it's not necessarily a given that concurrent access happens on the `provider`. It depends on timing. We could probably use a `CountDownLatch` to ensure that concurrent access happens at the right time.